File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
N, M, V = map(int, input().split())
V -= 1
EDGES = [[] for _ in range(N)]

for _ in range(M):
    a, b = map(lambda x: int(x) - 1, input().split())
    EDGES[a].append(b)
    EDGES[b].append(a)
    EDGES[a].sort()
    EDGES[b].sort()

def dfs():
    visited = [False for _ in range(N)]
    def _dfs(v, visited):
        visited[v] = True
        print(v + 1, end=' ')
        for edge in EDGES[v]:
            if not visited[edge]:
                _dfs(edge, visited)
    _dfs(V, visited)
    print()


def bfs():
    visited = [False for _ in range(N)]
    visited[V] = True

    def _bfs(start, visited):
        from collections import deque

        queue = deque([start])
        while queue:
            v = queue.popleft()
            print(v + 1, end=' ')
            for edge in EDGES[v]:
                if not visited[edge]:
                    queue.append(edge)
                    visited[edge] = True

    _bfs(V, visited)
    print()

dfs()
bfs()

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int V = Integer.parseInt(st.nextToken());

        List<Integer>[] EDGES = new ArrayList[N];
        for (int i = 0; i < N; i++) {
            EDGES[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            EDGES[s - 1].add(v - 1);
            EDGES[v - 1].add(s - 1);

            Collections.sort(EDGES[s - 1]);
            Collections.sort(EDGES[v - 1]);
        }

        new DFS(N, EDGES).recursive(V - 1);
        System.out.println();
        bfs(N, V, EDGES);
    }

    static class DFS {
        List<Integer>[] EDGES;
        boolean[] visit;

        DFS(int N, List<Integer>[] EDGES) {
            this.visit = new boolean[N];
            this.EDGES = EDGES;
        }

        public void recursive(int V) {
            if (!visit[V]) {
                System.out.print((V + 1) + " ");
                visit[V] = true;
            }
            for (int i = 0; i < EDGES[V].size();i++) {
                if (!visit[EDGES[V].get(i)]) recursive(EDGES[V].get(i));
            }
        }
    }

    public static void bfs(int N, int V, List<Integer>[] EDGES) {
        StringBuffer sb = new StringBuffer();

        Queue<Integer> queue = new ArrayDeque<>();
        queue.offer(V - 1);

        boolean[] visit = new boolean[N];
        visit[V - 1] = true;

        while (!queue.isEmpty()) {
            Integer pop = queue.poll();
            sb.append(pop + 1).append(" ");

            for (int i = 0; i < EDGES[pop].size(); i++) {
                int idx = EDGES[pop].get(i);
                if (!visit[idx]) {
                    visit[idx] = true;
                    queue.offer(idx);
                }
            }
        }

        System.out.println(sb);
    }
}